### PR TITLE
fix: disallow `#[test]` on associated functions

### LINKED ISSUE #6435: Disallow applying the `#[test]` attribute to impl methods
We recently had an issue in the bignum library not registering that we had multiple tests with the same name due to the tests being defined inside of an impl (for "hacky forcing the compiler to do things it doesn't want to do" reasons).

I don't see a good reason to have tests be anything other than free functions so we should error in cases where they're attached to impls.

### PR DESCRIPTION
**Pull request by asterite**
# Description

## Problem

Resolves #6435

## Documentation

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/hir/def_collector/errors.rs [hunk 1]
```diff
@@ -291,6 +293,12 @@ impl<'a> From<&'a DefCollectorErrorKind> for Diagnostic {
                 diag
             }
             DefCollectorErrorKind::UnsupportedNumericGenericType(err) => err.into(),
+            DefCollectorErrorKind::TestOnAssociatedFunction { span } => Diagnostic::simple_error(
+                "The `#[test]` attribute may only be used on a non-associated function".into(),
```
Comment (michaeljklein) [inline compiler/noirc_frontend/src/hir/def_collector/errors.rs:297]: What's a "non-associated function"? Closures are a sort of function too, but they don't allow `#[test]` either.

```suggestion
                "The `#[test]` attribute is disallowed on `impl` methods".into(),
```
?

**Review by michaeljklein — CHANGES_REQUESTED** [review]

File: compiler/noirc_frontend/src/hir/def_collector/errors.rs (hunk 1 above)
Comment (asterite) [reply compiler/noirc_frontend/src/hir/def_collector/errors.rs:297]: Sounds good. I copied the message from Rust 😊

**Review by jfecher — APPROVED** [review]

**Comment by asterite** [comment]
@TomAFrench Let me know if this can be merged. I think with the current macros we can workaround how it's done in the bignum library.

**Review by michaeljklein — APPROVED** [review]
LGTM

**Review by TomAFrench — APPROVED** [review]

**Comment by TomAFrench** [comment]
Thanks!

**Pull request merged**